### PR TITLE
Add cohere asr

### PR DESCRIPTION
**Pull request by eustlb**
# What does this PR do?
 
**Integration notes:**
For now, this integration does not load mel filters from the checkpoint. The original model was trained backpropagating gradients in it, but we saw previously (with parakeet-ctc) that this does not affect performance much. 

**Benchmarks (WER):**
| Dataset      | Orig  | Trfms (librosa f32) | Trfms (orig filters) |
|--------------|-------|---------------------|----------------------|
| AMI          | 8.16% | 8.16%               | 8.16%                |
| LibriSpeech  | 1.24% | 1.24%               | 1.25%                |
| FLEURS en    | 5.61% | 5.57%               | 5.61%                |
| FLEURS fr    | 4.76% | 4.81%               | 4.75%                |
| FLEURS de    | 4.15% | 4.19%               | 4.15%                |
| TED-LIUM     | 2.22% | 2.25%               | 2.23%                |

### CODE REVIEW

File: src/transformers/models/cohere_asr/configuration_cohere_asr.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+from ...configuration_utils import PreTrainedConfig
+from ...modeling_rope_utils import RopeParameters
+from ...utils import auto_docstring
+from ..auto import CONFIG_MAPPING
+from ..parakeet.configuration_parakeet import ParakeetEncoderConfig
+
+
+@auto_docstring(checkpoint="TODO")
+@strict(accept_kwargs=True)
```
Comment (ArthurZucker) [inline src/transformers/models/cohere_asr/configuration_cohere_asr.py]:
```suggestion
@strict
```

File: src/transformers/models/cohere_asr/processing_cohere_asr.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            raise ValueError("`language` must be provided when `audio_chunk_index` is given.")
+        separator = "" if language in _NO_SPACE_LANGS else " "
+        return self._reassemble_chunk_texts(texts, audio_chunk_index, separator)
+
+    @staticmethod
+    def _reassemble_chunk_texts(
+        texts: list[str],
+        audio_chunk_index: list[tuple[int, int | None]],
+        separator: str = " ",
+    ) -> list[str]:
```
Comment (ArthurZucker) [inline src/transformers/models/cohere_asr/processing_cohere_asr.py:137]: to document

**Review by ArthurZucker — APPROVED** [review]

File: docs/source/en/model_doc/cohere_asr.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+## Usage
+
+### Short-form transcription
+
+```python
+from transformers import AutoProcessor, CohereAsrForConditionalGeneration
+from transformers.audio_utils import load_audio
+
+processor = AutoProcessor.from_pretrained("cohere-ai/cohere-asr")
+model = CohereAsrForConditionalGeneration.from_pretrained("cohere-ai/cohere-asr", device_map="auto")
```
Comment (xenova) [inline docs/source/en/model_doc/cohere_asr.md]: `CohereLabs/cohere-transcribe-03-2026` :D

Comment (xenova) [reply docs/source/en/model_doc/cohere_asr.md]: and as it stands, the sample code has some issues complaining we need trust remote code due to custom tokenization

Comment (eustlb) [reply docs/source/en/model_doc/cohere_asr.md]: yes this was not up to date ahah! always finishing with the doc 😅

File: src/transformers/models/cohere_asr/modular_cohere_asr.py [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+    MoonshineDecoderMLP,
+    MoonshineForConditionalGeneration,
+    MoonshineModel,
+    MoonshinePreTrainedModel,
+    eager_attention_forward,
+)
+from .configuration_cohere_asr import CohereAsrConfig
+
+
+class CohereAsrDecoderMLP(MoonshineDecoderMLP):
```
Comment (vasqu) [inline src/transformers/models/cohere_asr/modular_cohere_asr.py]: CLIP MLP should be a drop in replacement

File: src/transformers/models/cohere_asr/modular_cohere_asr.py [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+        self.fc2 = nn.Linear(config.intermediate_size, config.hidden_size)
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        hidden_states = self.fc1(hidden_states)
+        hidden_states = self.activation_fn(hidden_states)
+        hidden_states = self.fc2(hidden_states)
+        return hidden_states
+
+
+class CohereAsrSelfAttention(nn.Module):
```
Comment (vasqu) [inline src/transformers/models/cohere_asr/modular_cohere_asr.py:53]: Can still inherit for the init only?

**Comment by eustlb** [comment]
@ArthurZucker the remaining failing is unrelated, ready to merge !! 😁

File: src/transformers/models/cohere_asr/modular_cohere_asr.py [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+            value_states,
+            attention_mask,
+            dropout=0.0 if not self.training else self.attention_dropout,
+            scaling=self.scaling,
+            **kwargs,
+        )
+
+        attn_output = attn_output.reshape(*input_shape, -1).contiguous()
+        attn_output = self.o_proj(attn_output)
+        return attn_output, attn_weights
```
Comment (vasqu) [inline src/transformers/models/cohere_asr/modular_cohere_asr.py:195]: Imo, we can reuse bert cross attn (needs override still because of o proj): this way we don't reimplement the same cache handling a different way again

File: src/transformers/models/cohere_asr/modular_cohere_asr.py [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+
+
+class CohereAsrCrossAttention(nn.Module):
```
Comment (vasqu) [inline src/transformers/models/cohere_asr/modular_cohere_asr.py:120]: Can also inherit for the init only

File: src/transformers/models/cohere_asr/modular_cohere_asr.py (hunk 6 above)
Comment (vasqu) [reply src/transformers/models/cohere_asr/modular_cohere_asr.py:195]: https://github.com/huggingface/transformers/blob/e3f7cc3b4a5499c52ad56c72cad295de562f5b54/src/transformers/models/bert/modeling_bert.py#L234

File: src/transformers/models/cohere_asr/modular_cohere_asr.py [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+            config.hidden_size, config.num_key_value_heads * self.head_dim, bias=config.attention_bias
+        )
+        self.o_proj = nn.Linear(
+            config.num_attention_heads * self.head_dim, config.hidden_size, bias=config.attention_bias
+        )
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        cross_attention_states: torch.Tensor | None = None,
```
Comment (vasqu) [inline src/transformers/models/cohere_asr/modular_cohere_asr.py]:
```suggestion
        encoder_hidden_states: torch.Tensor | None = None,
```
let's align with the other models, dont want another convention 😢

File: src/transformers/models/cohere_asr/modular_cohere_asr.py [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+        encoder_attention_mask = create_bidirectional_mask(
+            config=self.config,
+            inputs_embeds=inputs_embeds,
+            attention_mask=encoder_attention_mask,
+            encoder_hidden_states=encoder_hidden_states,
+        )
+
+        hidden_states = inputs_embeds
+        # No RoPE — position info is already in the embeddings via fixed sinusoidal pos_emb
+        position_embeddings = None
```
Comment (vasqu) [inline src/transformers/models/cohere_asr/modular_cohere_asr.py]: Why do we keep it in the signatures then?

**Review by vasqu — COMMENTED** [review]
Some quick comments for aligning a bit with modular maybe

File: src/transformers/models/cohere_asr/modular_cohere_asr.py (hunk 8 above)
Comment (eustlb) [reply src/transformers/models/cohere_asr/modular_cohere_asr.py]: went for this to match blt's 🫠

**Review by vasqu — APPROVED** [review]
Lets go
